### DISCUSSION TITLE
Add up numbers in item description

### DISCUSSION THREAD

**Discussion by Belgadon** [TomBursch/kitchenowl#89] [Ideas]
I would like to hear your opinion on summing up numbers in item descriptions.

There are currently two different behaviors when adding item with the same name multiple times.

**The items added have a description:**
Items are merged to one item and a comma separated list of descriptions is added.

**The items added _don't_ have a description:**
Items are merged to one item and and "..." is added as description. ("..." is added only once")

![liste2](https://user-images.githubusercontent.com/36801399/200332554-00baff75-7d91-4133-97c6-02004358cfb0.PNG)

What do you think of adding a feature to add up the numbers in the description field?
Should we just blindly add up all numbers?
Or only when there are only numbers in the description but no other characters?
Maybe also numbers with standard units attached to a number (like 500g)?
What about items without a description? Just do +1?

I'd add numbers in description and try to keep units: 500g + 500g = 1.000g
For items without description (or without a number) I would assume +1.
Resulting in _Kartoffeln (2.500g)_, _Parmesan (5)_, and _Zitrone (5)_ in my example above.

What about more complicated cases:
More than one number in the description, like "3 cans a 400g".
For me personal I would sanitize my recipe to either "3 cans" or "1.200g" to avoid problems like this.
Still, we should think about cases like this.

OR do we need a new filed "amount" next to description?

**Comment by irishrain** [comment]
I always thought the "description" field was misnamed, as it actually is meant to be used for "amount". Or do you think people are using it for things like "big", "from $Brand", "canned"?

Would you make the proposed "amount" field numerical? Or add a unit as well?

If the goal is to keep the UI lean, I think my intuitive expectation would be that if the descriptions in two recipes both start with a number and are then followed by the same string (which can contain more numbers - as in the "3 cans a 400g" example above) KitchenOwl would add the numbers. If the format differs, just attach as a list.

So "30 ml", "90 ml", "3 cans a 400 ml" and "2 cans a 400 ml" would give "120 ml, 5 cans a 400 ml".

I'm really unsure about things that don't have a number. The current "..." solution reminds you to check what you wanted to use it for. Addings the numbers might often be right, but often not - imagine 3 recepies with just "Kartoffeln". If this is summed up to 3 you might buy 3 single ones instead of the intended 3 bags. Maybe the solution would be to add " times", so the display would become "3 times", prompting you to check what 3 recipies you are needing it for.

**Reply by Belgadon** [reply]
> I always thought the "description" field was misnamed, as it actually is meant to be used for "amount". Or do you think people are using it for things like "big", "from $Brand", "canned"?

That makes perfect sense. In my opinion renaming it to amount would be a good choice.
Brand names should go into the name field.

> Would you make the proposed "amount" field numerical? Or add a unit as well?

I would also allow a String. e.g. "3ml olive oil" or "3 tablespoons/cups olive oil" can make a difference.

> If the goal is to keep the UI lean, I think my intuitive expectation would be that if the descriptions in two recipes both start with a number and are then followed by the same string (which can contain more numbers - as in the "3 cans a 400g" example above) KitchenOwl would add the numbers. If the format differs, just attach as a list.

>So "30 ml", "90 ml", "3 cans a 400 ml" and "2 cans a 400 ml" would give "120 ml, 5 cans a 400 ml".

Very good suggestion. Nothing gets lost, and everything is summarized when possible.
Things like "1 can a 400ml" and "2 can**_s_** a 400ml" would still be seperated, but I guess thats fine. The user can still clean the recipe.

> I'm really unsure about things that don't have a number. The current "..." solution reminds you to check what you wanted to use it for. Addings the numbers might often be right, but often not - imagine 3 recepies with just "Kartoffeln". If this is summed up to 3 you might buy 3 single ones instead of the intended 3 bags. Maybe the solution would be to add " times", so the display would become "3 times", prompting you to check what 3 recipies you are needing it for.

I really like that "3 times" ("3x"?) proposal. Makes sence for for a lot of use cases.
[I just realized that an option to add items without a recipe would be useful. Like milk, cereal or bread.]

**Reply by Belgadon** [reply]
The same logic should apply when changing the servings before adding it to the shopping list.
Right now i adds up all numbers "3 cans a 400ml" -> "6 cans a 800ml".
Items without description aren't changet at all. (no "...")

**Reply by TomBursch** [reply]
So finally got around to give my two cents here :) 
> I always thought the "description" field was misnamed, as it actually is meant to be used for "amount". Or do you think people are using it for things like "big", "from $Brand", "canned"?

Interestingly that is what I'm sometimes doing. Especially if I don't plan to buy that item/brand often. So I have an item `Sliced cheese` with an description of `Gouda` for example. This allows to use the existing algorithmic ordering/categories for that item. 

> If the goal is to keep the UI lean

That is definitely a goal, I don't want to have many different settings related to that. I don't want something like with grocery where you have to manage units and the conversion rates between them. 

Like you both said, I expect everyone to somehow sanitize their recipes and not have something like `3 cans a 400ml` then a system that detects just basic SI units and add them together could work. In cases where the unit could not be determined I would just append that to the end of the list.

For example merging `300ml` + `1` + `1L` + `2` + ` 2 halves` should result in `1300ml, 3, 2 halves`. This could even be extended to detect the same non existing unit, so adding `3 halves` to that would yield in `1300ml, 3, 5 halves`.
Everything else which doesn't contain a number could just be added to the end. E.g. adding `Gouda` results in `1300ml, 3, 5 halves, Gouda`.

**Reply by irishrain** [reply]
What would you do when something without a description is added?

**Reply by TomBursch** [reply]
I guess I like both approaches. Like it has been said `...` tells you to check the planned recipes and having number x times tells you exactly how many recipes it was used in. But it sometimes does not correlate to how much is actually needed. Say I have two recipes with vegetable broth and haven't specified how much is needed because if I put that on the list it usually means just by one glass. Then `...` just tells me it is needed for multiple recipes whereas 2x would suggest I need two glasses. Even though I should still only buy one.

Another problem that I thought of when processing amounts, and the same issue persists for servings, is how different languages handle decimal marks. `1,000g` could be either 1g or 1000g depending on the input language. I couldn't figure out a good solution to that problem. When scaling a recipe I just replace all `,` with `.`. Although this will cause issues in English for example.

**Reply by irishrain** [reply]
>  [I just realized that an option to add items without a recipe would be useful. Like milk, cereal or bread.]

Just type it into the search field on the shopping list - or am I misunderstanding what you mean?

**Reply by Belgadon** [reply]
### With description:

> `30 ml`, `90 ml`, `3 cans a 400 ml` and `2 cans a 400 ml` would give `120 ml, 5 cans a 400 ml`

>For example merging `300ml` + `1` + `1L` + `2` + `2 halves` should result in `1300ml, 3, 2 halves`. This could even be extended to detect the same non existing unit, so adding `3 halves` to that would yield in `1300ml, 3, 5 halves`.
Everything else which doesn't contain a number could just be added to the end. E.g. adding `Gouda` results in `1300ml, 3, 5 halves, Gouda`.

This would be a good first step.
Using SI units would be even better but not necessary imho.

> Another problem that I thought of when processing amounts, and the same issue persists for servings, is how different languages handle decimal marks. `1,000g` could be either 1g or 1000g depending on the input language. I couldn't figure out a good solution to that problem. When scaling a recipe I just replace all `,` with `.`. Although this will cause issues in English for example.

I think one could check for localization of used device and change the behavior accordingly, right?
Or have a number format option in settings?

### Without description:
> I guess I like both approaches. Like it has been said `...` tells you to check the planned recipes and having number x times tells you exactly how many recipes it was used in. But it sometimes does not correlate to how much is actually needed. Say I have two recipes with vegetable broth and haven't specified how much is needed because if I put that on the list it usually means just by one glass. Then `...` just tells me it is needed for multiple recipes whereas 2x would suggest I need two glasses. Even though I should still only buy one.

With `...` you don't see in how many or which recipes you have to look, or e.g. for how many people (cook recipes X twice cause you get visitors on Friday).
Checking that while grocery shopping could be time consuming and frustrating.

I think `3x` would be more useful than `...`. Even if you have no clue in that moment you could probably check the "planer" tab and instantly know "right, it was once for recipes A and I wanted to do recipes B twice (more people), two glasses should be enough (not 3).

> Addings the numbers might often be right, but often not - imagine 3 recepies with just `Kartoffeln`. If this is summed up to 3 you might buy 3 single ones instead of the intended 3 bags.

The recipes should contain `1 bag of Kartoffeln`. Otherwise, when adding ONLY this recipes  you still wouldn't know how much to buy. :-)

At the end, I'd doubt that there is a solution for every case. You still always have to use your brain and one could always clean up the recipes beforehand.

**Comment by TomBursch** [comment]
This is now implemented with #100 

Feel free to comment if you notice any strange behavior.
Things that currently are a bit strange:
Ranges: `"1-2" + "3-4"` results in `1 -2, 3 -4`
Amounts with description: `"100g fresh" + "100g fresh"` results in: `200 g fresh` and SI unit merging won't work.

Everything else should be as expected, including merging amount of SI units.
You can find input output examples in the [test](https://github.com/TomBursch/kitchenowl-backend/blob/23a863be72d8fa1926aae03076a500550bf89282/tests/util/test_description_merger.py#L5)
